### PR TITLE
Changes needed for SUSY parameter scans

### PR DESCRIPTION
**Pull request by bendavid**
Adds the possibility to provide a list of configurations to Pythia8 and randomly select from the list of possible configurations at beginLuminosityBlock time.

This allows to produce a parameter scan in a single monte carlo request/dataset for whatever can be varied at the pythia level. This is targeted initially for SUSY fastsim signal production in 80x where the parameter scan is over the SUSY decay chain defined by the SLHA header, which can be inlined as part of the pythia8 configuration.

Since the changes are non-trivial agreed to put it in 81x first and backport afterwards.

This is the (near) final implementation of the proposal discussed in https://indico.cern.ch/event/487292/contribution/15/attachments/1218978/1781280/susyscan-Jan28-2016.pdf

Doing this safely requires a more complete re-initizalization of pythia at lumi section boundaries (pointer to main program and those related to userhooks are completely deleted and recreated rather than just calling init() again)

I've tested that this does not leak memory at any observable rate over several thousand lumi sections.

The inheritance of the Pythia8 interfaces was also slightly reorganized so that everything really inherits from BaseHadronizer now. (previously some of the particle guns did not, which made it more painful to extend the hadronizer api)

The DataFormats changes are needed to store in the event information on which config has been chosen as well as the LHE-related info discussed below.  A new GenLumiInfoHeader class is created so that it can be filled in beginLuminosityBlock and accessed within a lumi section if needed.  (existing GenLumiInfoProduct needs information only available at the end of the lumi section)

There is an important new functionality on top of this to allow accessing a gridpack and producing LHE events within the BaseHadronizer at beginLuminosityBlock time instead of in the ExternalLHEProducer.  In this case gridpacks can also be randomly selected together with the pythia configuration.  LHE events produced in this way bypass completely the existing LHE machinery in CMSSW and use directly the Pythia8 LHE parser.  In this case the related information on additional weights must be stored in the GenEventInfoProduct and new GenLumiInfoHeader.

Note that due to some issues in the Pythia8 LHE parser the information related to which weight is which in the GenEventInfoProduct is currently incomplete.  This will require some further patches and/or updates to pythia8 to fully resolve, but this pull request is already at least functional without it, so we should get it into the release and tested.

This pull request replaces https://github.com/cms-sw/cmssw/pull/13433

### CODE REVIEW

**Comment by bendavid** [comment]
One further remaining issue here is that the GenXSecAnalyzer printout should be improved for the case where LHE's are produced in BaseHadronizer.  (cross section should already be correct, though needs to be checked for weighted events, but currently matching efficiency is not clearly indicated by the printout)

File: GeneratorInterface/Core/src/BaseHadronizer.cc [hunk 1]
```diff
@@ -1,20 +1,154 @@
...
+    return genLumiInfoHeader;
+    
+  }
 
-}
+  void BaseHadronizer::randomizeIndex(edm::LuminosityBlock const& lumi, CLHEP::HepRandomEngine* rengine) {
+    if (randomInitWeights_.size()>0) {
+      //randomly select from a list of provided configuration sets (for parameter scans)
+      
+      //seeds std 32-bit mersene twister with HepRandomEngine state plus run and lumi section numbers
```
Comment (bendavid) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: @Dr15Jones this is the really final iteration of the random seed initialization where now literally the full internal state of the HepJamesRandom is used together with the run and lumi numbers to initialize the std random number generator.

File: GeneratorInterface/Core/src/BaseHadronizer.cc [hunk 2]
```diff
@@ -1,20 +1,154 @@
...
+    }
+  }
+  
+  void BaseHadronizer::generateLHE(edm::LuminosityBlock const& lumi, CLHEP::HepRandomEngine* rengine) {
+        
+    if (gridpackPath().empty()) {
+      return;
+    }
+    
+    //get random seed from HepRandomEngine state plus run and lumi section numbers
```
Comment (bendavid) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: and another variant of the random number initialization here where in this case no random number generator is seeded within the code, but the HepJamesRandom state is used with run and lumi numbers to generate a new (single) seed for the gridpack lhe generation script.

**Comment by bendavid** [comment]
For what concerns event size, the additional lumi product is about 13kB/lumi section in the worst case scenario (LO madgraph with several hundred weights)

File randomizedParametersSLHALHE_py_GEN.root Events 3
Branch Name | Average Uncompressed Size (Bytes/Event) | Average Compressed Size (Bytes/Event) 
GenLumiInfoProduct_generator__GEN. 17875.7 17875.7
GenLumiInfoHeader_generator__GEN. 58514 12638.7
GenFilterInfo_genFilterEfficiencyProducer__GEN. 6475.33 6475.33
LuminosityBlockAuxiliary 370.667 186

The GenEventInfoProduct with many weights gets largish, but this extra space is ~exactly equivalent to what is saved on account of the LHEEventInfoProduct not being present.
GenEventInfoProduct_generator__GEN. 4844.06 4182.22

Would be nice to switch from double to float for the GenEventInfoProduct weights, but this would be a  dataformat change in a pretty critical place affecting potentially all samples and analysis workflows, so avoided for now.

File: GeneratorInterface/Core/src/BaseHadronizer.cc (hunk 1 above)
Comment (Dr15Jones) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: @wddgit could you do a once over on the random number use?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by bendavid** [comment]
@cmsbuild please test

**Comment by bendavid** [comment]
Previous problem should be fixed now.  (Code was not checking for null pointer and would crash in case lhe information was not present in the pythia Info object, now it should just skip filling the missing information in that case.)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13530	90e1fd7
- code changes in reco packages are minimal: addition of  ``` GenLumiInfoHeader_generator_ ``` to the miniAOD event content
- jenkins tests pass and comparisons show no differences; checking jenkins output files (wf 1330 step1), there are no new products added in MC workflow(s) and existing products are unchanged.

**Comment by bendavid** [comment]
@slava77 there SHOULD be a new product (the GenLumiInfoHeader) added in the MC workflows (but only the in the luminosityBlocks branch)

File: GeneratorInterface/Core/src/BaseHadronizer.cc [hunk 3]
```diff
@@ -1,20 +1,154 @@
...
+      std::vector<long unsigned int> seeds = rengine->put();
+      seeds.push_back(lumi.id().run());
+      seeds.push_back(lumi.id().luminosityBlock());
+      std::seed_seq seedseq(seeds.begin(),seeds.end());
+      std::mt19937 randgen(seedseq);
+      std::discrete_distribution<int> randdist(randomInitWeights_.begin(),randomInitWeights_.end());
+      
+      randomIndex_ = randdist(randgen);
+      
+      printf("selected random index = %i\n",randomIndex_);
```
Comment (covarell) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: This seems to go to stdout. Should it be removed?

**Comment by slava77** [comment]
@bendavid 
My original check was in the Events.
I looked now in LuminosityBlocks and I see GenLumiInfoHeader_generator added.
In this case 12kB is in GenLumiInfoHeader, compared to 45 kB in GenLumiInfoProduct

File: GeneratorInterface/Core/src/BaseHadronizer.cc (hunk 3 above)
Comment (bendavid) [inline GeneratorInterface/Core/src/BaseHadronizer.cc]: yup, missed that one, sorry.  Will take it out.

**Comment by covarell** [comment]
Other question is why all the methods removed to Py8ParticleGun (is particle gun connected with this PR?)

**Comment by bendavid** [comment]
The changes to particle gun are due to a cleanup of the inheritance structure.
Before this pull request, the pythia8 particle guns did not inherit from BaseHadronizer.  (and were the only hadronizers in the release not to do so)

This made it painful to extend the hadronizer interface, because additional functions would need to be explicitly implemented for the particle guns.  So the inheritance was rearranged in the pythia8 interface such that all hadronizers now inherit from BaseHadronizer (at least indirectly)

**Comment by slava77** [comment]
+1

for 06c3c39
-  GeneratorInterface/Core/src/BaseHadronizer.cc  change only, no effect on reco

**Comment by bendavid** [comment]
(so the functions which were removed from the guns are no longer needed because they are already implemented by some of the base classes)

**Comment by covarell** [comment]
+1

**Comment by bendavid** [comment]
Anything still holding this up?  Analysis signature is needed for miniaod event content change?

**Comment by davidlange6** [comment]
just time for me to look at it… sorry for the holdup

**Comment by davidlange6** [comment]
@Dr15Jones @wddgit - were you ok with the random number implementation in the end?

**Comment by wddgit** [comment]
This looks OK to me now. I cannot see any problems.

**Comment by davidlange6** [comment]
+1

**Pull request merged**